Tests/GRPCNIOTransportHTTP2Tests/HTTP2TransportTLSEnabledTests.swift

```diff
@@ -0,0 +1,444 @@
+      await #expect(
+        performing: {
+          try await self.executeUnaryRPC(control: control)
+        },
+        throws: { error in
+          guard let rootError = error as? RPCError else {
+            Issue.record("Should be an RPC error")
+            return false
+          }
+          #expect(rootError.code == .unavailable)
+          #expect(
+            rootError.message
+              == "The server accepted the TCP connection but closed the connection before completing the HTTP/2 connection preface."
+          )
+
+          guard
+            let sslError = rootError.cause as? NIOSSLExtraError,
+            case .failedToValidateHostname = sslError
+          else {
+            Issue.record(
+              "Should be a NIOSSLExtraError.failedToValidateHostname error, but was: \(String(describing: rootError.cause))"
+            )
+            return false
+          }
+
+          return true
+        }
+      )
+    }
+  }
+
+  @Test(
+    "Error is surfaced when server fails client verification",
+    arguments: [TransportSecurity.posix],
+    [TransportSecurity.posix]
+  )
+  // Verification should fail because the server does not have trust roots containing the client cert.
+  func testServerFailsClientValidation(
+    clientTransport: TransportSecurity,
+    serverTransport: TransportSecurity
+  ) async throws {
+    let certificateKeyPairs = try SelfSignedCertificateKeyPairs()
+    let clientTransportConfig = self.makeMTLSClientTLSConfig(
+      for: clientTransport,
+      certificateKeyPairs: certificateKeyPairs,
+      serverHostname: "localhost"
+    )
+    let serverTransportConfig = self.makeMTLSServerTLSConfig(
+      for: serverTransport,
+      certificateKeyPairs: certificateKeyPairs,
+      includeClientCertificateInTrustRoots: false
+    )
+
+    try await self.withClientAndServer(
+      clientTransportSecurity: clientTransportConfig,
+      serverTransportSecurity: serverTransportConfig
+    ) { control in
+      await #expect(
+        performing: {
+          try await self.executeUnaryRPC(control: control)
+        },
+        throws: { error in
+          guard let rootError = error as? RPCError else {
+            Issue.record("Should be an RPC error")
+            return false
+          }
+          #expect(rootError.code == .unavailable)
+          #expect(
+            rootError.message
+              == "The server accepted the TCP connection but closed the connection before completing the HTTP/2 connection preface."
+          )
+
+          guard
+            let sslError = rootError.cause as? NIOSSL.BoringSSLError,
+            case .sslError = sslError
+          else {
+            Issue.record(
+              "Should be a NIOSSL.sslError error, but was: \(String(describing: rootError.cause))"
+            )
+            return false
+          }
+
+          return true
+        }
+      )
+    }
+  }
+
+  // - MARK: Test Utilities
+
+  enum TransportSecurity: Sendable {
```

Comment:
`TransportKind` is probably a more appropriate name here, security would be TLS/plaintext